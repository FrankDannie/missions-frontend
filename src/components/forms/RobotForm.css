@import './Form.css';

/* Styling for the robot form container */
.robot-form {
  /* The .form-container class from Form.css will handle this */
}

.robot-form h2 {
  margin-bottom: 20px; /* Space below the heading */
  font-size: 1.5rem; /* Increase heading size for visibility */
}

.robot-form label {
  display: block; /* Display label as block element for full-width */
  margin-bottom: 10px; /* Space below label */
  font-weight: bold; /* Bold text for labels */
}

.robot-form input,
.robot-form textarea {
  width: 100%; /* Full-width input and textarea */
  padding: 10px; /* Increased padding for better usability */
  margin-top: 5px; /* Space above input and textarea */
  border: 1px solid #ccc; /* Light border color */
  border-radius: 4px; /* Rounded corners */
  box-sizing: border-box; /* Include padding and border in width */
}

.robot-form textarea {
  resize: vertical; /* Allow vertical resizing of textareas */
  min-height: 100px; /* Minimum height for textareas */
}

.robot-form button[type='submit'],
.robot-form button[type='button'] {
  padding: 10px 20px; /* Consistent padding for buttons */
  border: none; /* Remove default button border */
  border-radius: 4px; /* Rounded corners */
  cursor: pointer; /* Pointer cursor on hover */
  font-size: 1rem; /* Consistent font size */
  transition: background-color 0.3s ease; /* Smooth background color transition */
}

.robot-form button[type='submit'] {
  background-color: #007bff; /* Primary color for submit button */
  color: #fff; /* White text */
}

.robot-form button[type='submit']:hover {
  background-color: #0056b3; /* Darker shade on hover */
}

.robot-form button[type='button'] {
  background-color: #6c757d; /* Secondary color for button */
  color: #fff; /* White text */
}

.robot-form button[type='button']:hover {
  background-color: #5a6268; /* Darker shade on hover */
}

@media (max-width: 768px) {
  .robot-form h2 {
    font-size: 1.25rem; /* Slightly smaller heading on smaller screens */
  }

  .robot-form input,
  .robot-form textarea {
    padding: 8px; /* Adjust padding for smaller screens */
  }

  .robot-form button {
    padding: 8px 15px; /* Adjust button padding for smaller screens */
  }
}
