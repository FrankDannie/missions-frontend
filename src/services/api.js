import axios from 'axios';

const api = axios.create({
  baseURL: process.env.REACT_APP_API_BASE_URL,
});

export const fetchRobots = () => api.get('/robots/');
export const fetchRobotById = (id) => api.get(`/robots/${id}/`); 
export const createRobot = (data) => api.post('/robots/', data);
export const updateRobot = (id, data) => api.put(`/robots/${id}/`, data);
