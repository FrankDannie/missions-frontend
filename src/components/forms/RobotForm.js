import React, { useState, useEffect } from 'react';
import { createRobot, updateRobot, fetchRobotById } from '../../services/api';
import './RobotForm.css';

const RobotForm = ({ robotId, onClose, onSubmit }) => {
  const [robot, setRobot] = useState({ name: '', model_name: '' });

  useEffect(() => {
    if (robotId) {
      fetchRobotById(robotId).then(response => setRobot(response.data));
    } else {
      setRobot({ name: '', model_name: '' });
    }
  }, [robotId]);

  const handleChange = (e) => {
    const { name, value } = e.target;
    setRobot(prevRobot => ({ ...prevRobot, [name]: value }));
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    if (robotId) {
      updateRobot(robotId, robot).then(() => {
        if (onSubmit) onSubmit(); // Trigger the submission actions
        setRobot({ name: '', model_name: '' }); // Reset the form fields
        onClose();
      });
    } else {
      createRobot(robot).then(() => {
        if (onSubmit) onSubmit(); // Trigger the submission actions
        setRobot({ name: '', model_name: '' }); // Reset the form fields
        onClose();
      });
    }
  };

  const handleCancel = () => {
    setRobot({ name: '', model_name: '' }); // Reset the form fields
    onClose();
  };

  return (
    <div className="form-container robot-form">
      <h2>{robotId ? 'Update Robot' : 'Create Robot'}</h2>
      <form onSubmit={handleSubmit}>
        <label>Name:
          <input type="text" name="name" value={robot.name} onChange={handleChange} required />
        </label>
        <label>Model Name:
          <input type="text" name="model_name" value={robot.model_name} onChange={handleChange} required />
        </label>
        <button type="submit">{robotId ? 'Update' : 'Create'}</button>
        <button type="button" onClick={handleCancel}>Cancel</button>
      </form>
    </div>
  );
};

export default RobotForm;
