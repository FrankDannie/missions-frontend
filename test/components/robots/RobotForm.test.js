import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import RobotForm from '../../../src/components/forms/RobotForm';
import { createRobot, updateRobot, fetchRobotById } from '../../../src/services/api';
import '@testing-library/jest-dom/extend-expect';
import Labels from '../../../src/components/common/labels';

jest.mock('../../../src/services/api');

describe('RobotForm', () => {
    test('renders form for creating a new robot', () => {
        const { asFragment } = render(<RobotForm onClose={() => {}} />);
    
        expect(screen.getByText(Labels.robot.CREATE_TITLE)).toBeInTheDocument();
    
        // Select the first element matching the label
        expect(screen.getAllByLabelText(/Name/i)[0]).toBeInTheDocument();
        expect(screen.getAllByLabelText(/Model Name/i)[0]).toBeInTheDocument();
        expect(screen.getByRole('button', { name: Labels.form.CREATE })).toBeInTheDocument();
        expect(screen.getByRole('button', { name: Labels.CANCEL })).toBeInTheDocument();
    
        // Snapshot for creating a new robot
        expect(asFragment()).toMatchSnapshot();
    });
    
    
  
    test('renders form with robot data for editing', async () => {
      fetchRobotById.mockResolvedValue({ data: { name: 'Robot 1', model_name: 'Model 1' } });
      const { asFragment } = render(<RobotForm robotId="1" onClose={() => {}} />);
  
      await waitFor(() => {
        expect(screen.getByDisplayValue('Robot 1')).toBeInTheDocument();
        expect(screen.getByDisplayValue('Model 1')).toBeInTheDocument();
      });
  
      // Snapshot for editing an existing robot
      expect(asFragment()).toMatchSnapshot();
    });
  
    test('submits form data correctly for updating an existing robot', async () => {
        const handleClose = jest.fn();
        updateRobot.mockResolvedValue({ data: {} });
        fetchRobotById.mockResolvedValue({ data: { name: 'Robot 1', model_name: 'Model 1' } });
    
        // Destructure asFragment from render
        const { asFragment } = render(<RobotForm robotId="1" onClose={handleClose} />);
    
        // Wait for the form to populate
        await waitFor(() => {
            expect(screen.getAllByLabelText(/Name/i)[0]).toHaveValue('Robot 1');
            expect(screen.getAllByLabelText(/Model Name/i)[0]).toHaveValue('Model 1');
        });
    
        fireEvent.change(screen.getAllByLabelText(/Name/i)[0], { target: { value: 'Updated Robot' } });
        fireEvent.change(screen.getAllByLabelText(/Model Name/i)[0], { target: { value: 'Updated Model' } });
        fireEvent.click(screen.getByRole('button', { name: Labels.form.UPDATE }));
    
        await waitFor(() => {
            expect(updateRobot).toHaveBeenCalledWith('1', { name: 'Updated Robot', model_name: 'Updated Model' });
            expect(handleClose).toHaveBeenCalled();
        });
    
        expect(asFragment()).toMatchSnapshot();
    });
  });