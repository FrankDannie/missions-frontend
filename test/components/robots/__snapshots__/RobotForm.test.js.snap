// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RobotForm renders form for creating a new robot 1`] = `
<DocumentFragment>
  <div
    class="form-container robot-form"
  >
    <h2>
      Create Robot
    </h2>
    <form>
      <label>
        Name
        <input
          name="name"
          required=""
          type="text"
          value=""
        />
      </label>
      <label>
        Model Name
        <input
          name="model_name"
          required=""
          type="text"
          value=""
        />
      </label>
      <button
        type="submit"
      >
        Create
      </button>
      <button
        type="button"
      >
        Cancel
      </button>
    </form>
  </div>
</DocumentFragment>
`;

exports[`RobotForm renders form with robot data for editing 1`] = `
<DocumentFragment>
  <div
    class="form-container robot-form"
  >
    <h2>
      Update Robot
    </h2>
    <form>
      <label>
        Name
        <input
          name="name"
          required=""
          type="text"
          value="Robot 1"
        />
      </label>
      <label>
        Model Name
        <input
          name="model_name"
          required=""
          type="text"
          value="Model 1"
        />
      </label>
      <button
        type="submit"
      >
        Update
      </button>
      <button
        type="button"
      >
        Cancel
      </button>
    </form>
  </div>
</DocumentFragment>
`;

exports[`RobotForm submits form data correctly for updating an existing robot 1`] = `
<DocumentFragment>
  <div
    class="form-container robot-form"
  >
    <h2>
      Update Robot
    </h2>
    <form>
      <label>
        Name
        <input
          name="name"
          required=""
          type="text"
          value=""
        />
      </label>
      <label>
        Model Name
        <input
          name="model_name"
          required=""
          type="text"
          value=""
        />
      </label>
      <button
        type="submit"
      >
        Update
      </button>
      <button
        type="button"
      >
        Cancel
      </button>
    </form>
  </div>
</DocumentFragment>
`;
